flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -72,123 +197,184 @@ int32_t vtol_follower_control_path(const float dT, const PathDesiredData *pathDe
+		-guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
+
 	// Update the path status UAVO
 	PathStatusData pathStatus;
 	PathStatusGet(&pathStatus);
 	pathStatus.fractional_progress = progress->fractional_progress;
 	pathStatus.error = progress->error;
-	if (pathStatus.fractional_progress < 1)
+	pathStatus.Waypoint = pathDesired->Waypoint;
+
+	// Only if we are at the end, and above the end altitude, are we
+	// done.
+	if ((pathStatus.fractional_progress < 1) ||
+		(downError < -guidanceSettings.EndpointRadius)) {
 		pathStatus.Status = PATHSTATUS_STATUS_INPROGRESS;
-	else
+	} else {
 		pathStatus.Status = PATHSTATUS_STATUS_COMPLETED;
+	}
 	PathStatusSet(&pathStatus);
 
-	float groundspeed = pathDesired->StartingVelocity + 
-	    (pathDesired->EndingVelocity - pathDesired->StartingVelocity) * progress->fractional_progress;
-	if (progress->fractional_progress > 1)
-		groundspeed = 0;
-	
-	VelocityDesiredData velocityDesired;
-	velocityDesired.North = progress->path_direction[0] * groundspeed;
-	velocityDesired.East = progress->path_direction[1] * groundspeed;
-	
-	float error_speed = progress->error * 
+	if (groundspeed == 0) {
+		return vtol_follower_control_simple(dT, pathDesired->End, 0, 1);
+	}
+
+	float error_speed = vtol_deadband(progress->error,
+		guidanceSettings.PathDeadbandWidth,
+		guidanceSettings.PathDeadbandCenterGain,
+		vtol_path_m, vtol_path_r) *
 	    guidanceSettings.HorizontalPosPI[VTOLPATHFOLLOWERSETTINGS_HORIZONTALPOSPI_KP];
-	float correction_velocity[2] = {progress->correction_direction[0] * error_speed, 
-	    progress->correction_direction[1] * error_speed};
-	
-	float total_vel = sqrtf( powf(correction_velocity[0], 2) + powf(correction_velocity[1], 2) );
-	float scale = 1;
-	if (total_vel > guidanceSettings.HorizontalVelMax)
-		scale = guidanceSettings.HorizontalVelMax / total_vel;
-
-	// Currently not apply a PID loop for horizontal corrections
-	velocityDesired.North += progress->correction_direction[0] * error_speed * scale;
-	velocityDesired.East += progress->correction_direction[1] * error_speed * scale;
+
+	/* Sum the desired path movement vector with the correction vector */
+	commands_ned[0] = progress->path_direction[0] * groundspeed +
+	    progress->correction_direction[0] * error_speed;
 	
-	// Interpolate desired velocity along the path
-	float altitudeSetpoint = pathDesired->Start[2] + (pathDesired->End[2] - pathDesired->Start[2]) *
-	    bound_min_max(progress->fractional_progress, 0, 1);
+	commands_ned[1] = progress->path_direction[1] * groundspeed +
+	    progress->correction_direction[1] * error_speed;
 
-	float downError = altitudeSetpoint - positionActual.Down;
-	velocityDesired.Down = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
-		-guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
+	/* Limit the total velocity based on the configured value. */
+	vtol_limit_velocity(commands_ned, guidanceSettings.HorizontalVelMax);
```

Comment:
good catch. apply the limit to the requested velocity seems more correct than applying it to the error component only.